### ISSUE TITLE
Raft hiccup in TestRestoreReplicas

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#562]
The following test appears to have failed:

[#1113](https://circleci.com/gh/cockroachdb/cockroach/1113):

```
I0402 20:46:47.104527     257 multiraft.go:448] node 514: group 18446744073709551615 got message 257->0 MsgHeartbeatResp Term:0 Log:0/0
I0402 20:46:47.104636     257 multiraft.go:448] node 257: group 18446744073709551615 got message 514->257 MsgHeartbeat Term:0 Log:0/0
W0402 20:46:47.104687     257 multiraft.go:203] node 257: not fanning out heartbeat from unknown node 514 (but responding anyway)
I0402 20:46:47.104959     257 multiraft.go:448] node 257: group 18446744073709551615 got message 514->257 MsgHeartbeat Term:0 Log:0/0
W0402 20:46:47.105009     257 multiraft.go:203] node 257: not fanning out heartbeat from unknown node 514 (but responding anyway)
panic: test timed out after 5m0s

goroutine 662411 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x123f990, 0x16ed5c0, 0x8b, 0x8b, 0x1)
	/usr/src/go/src/testing/testing.go:556 +0xdc3
--
	/go/src/github.com/cockroachdb/cockroach/kv/local_sender.go:163 +0x1dd
github.com/cockroachdb/cockroach/kv.func·010()
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:133 +0x2a6
created by github.com/cockroachdb/cockroach/kv.(*txnMetadata).close
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:137 +0xbe0
FAIL	github.com/cockroachdb/cockroach/storage	300.074s
=== RUN TestBatchBasics
--- PASS: TestBatchBasics (0.01s)
=== RUN TestBatchGet
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
--- PASS: TestBatchMerge (0.00s)
=== RUN TestBatchProto
--- PASS: TestBatchProto (0.00s)
=== RUN TestBatchScan
--- PASS: TestBatchScan (0.01s)
I0402 20:46:47.104527     257 multiraft.go:448] node 514: group 18446744073709551615 got message 257->0 MsgHeartbeatResp Term:0 Log:0/0
I0402 20:46:47.104636     257 multiraft.go:448] node 257: group 18446744073709551615 got message 514->257 MsgHeartbeat Term:0 Log:0/0
W0402 20:46:47.104687     257 multiraft.go:203] node 257: not fanning out heartbeat from unknown node 514 (but responding anyway)
I0402 20:46:47.104959     257 multiraft.go:448] node 257: group 18446744073709551615 got message 514->257 MsgHeartbeat Term:0 Log:0/0
W0402 20:46:47.105009     257 multiraft.go:203] node 257: not fanning out heartbeat from unknown node 514 (but responding anyway)
panic: test timed out after 5m0s

goroutine 662411 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x123f990, 0x16ed5c0, 0x8b, 0x8b, 0x1)
	/usr/src/go/src/testing/testing.go:556 +0xdc3
--
	/go/src/github.com/cockroachdb/cockroach/kv/local_sender.go:163 +0x1dd
github.com/cockroachdb/cockroach/kv.func·010()
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:133 +0x2a6
created by github.com/cockroachdb/cockroach/kv.(*txnMetadata).close
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:137 +0xbe0
FAIL	github.com/cockroachdb/cockroach/storage	300.074s
=== RUN TestBatchBasics
--- PASS: TestBatchBasics (0.01s)
=== RUN TestBatchGet
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
--- PASS: TestBatchMerge (0.00s)
=== RUN TestBatchProto
--- PASS: TestBatchProto (0.00s)
=== RUN TestBatchScan
--- PASS: TestBatchScan (0.01s)

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
Something went wrong here: a replica change txn fails, and the test ends up in a configuration where one node knows the other (but not vice versa) and no progress is made except a perpetual interchange of heartbeats.

**Comment by tschottdorf** [comment]
hey @bdarnell, any pointers on this one?

**Comment by bdarnell** [comment]
I believe this was fixed by #573 (directly resolve intents in ChangeReplicas).

What happened was that the ChangeReplicas transaction completed, but intents were not resolved before the stores were stopped and restarted. On restart the stores read the pre-intent version of the range descriptor, so both replicas saw store 1 as the sole member of the range. Nothing ever resolves the write intent so the state of the range descriptor never moves forward to include both replicas. Now that we force resolution of the range descriptor intents the problem should be gone.
